File: backend/tiggle-root/tiggle-auth-gateway/src/main/java/com/side/tiggle/authgateway/oauth/JwtTokenProvider.java
package com.side.tiggle.authgateway.oauth;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.JwtException;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;
import org.springframework.http.HttpHeaders;
import org.springframework.http.server.reactive.ServerHttpRequest;
import org.springframework.stereotype.Component;

import javax.annotation.PostConstruct;
import javax.crypto.SecretKey;
import java.nio.charset.StandardCharsets;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.util.Date;

@Component
public class JwtTokenProvider {
    
    // TODO: application.yml에서 환경에 따라 가져오도록 설정해야 함
    private String secret = "secrettigglesecrettigglesecrettiggle";
    private SecretKey secretKey;

    private final long tokenExpiry = 60L * 30L; // sec
    private final long refreshTokenExpiry = 60L * 60L * 24L; // sec

    @PostConstruct
    void init(){
        secretKey = Keys.hmacShaKeyFor(secret.getBytes(StandardCharsets.UTF_8));
    }

    public String getAccessToken(long memberId, String roles) {
        return generateToken(memberId, roles, tokenExpiry);
    }

    public String getRefreshToken(long memberId, String roles) {
        return generateToken(memberId, roles, refreshTokenExpiry);
    }

    public String generateToken(long memberId, String roles, long validFor) {
        Date issued = Date.from(LocalDateTime.now().atZone(ZoneId.systemDefault()).toInstant());
        Date expiry = Date.from(LocalDateTime.now().plusSeconds(validFor).atZone(ZoneId.systemDefault()).toInstant());
        Claims claims = Jwts.claims().setSubject(String.valueOf(memberId));
        claims.put("roles", roles);

        return Jwts.builder()
                .setIssuer("tiggle")
                .setClaims(claims)
                .setIssuedAt(issued)
                .setExpiration(expiry)
                .signWith(secretKey)
                .compact();
    }

    public String resolveAccessToken(ServerHttpRequest request) {
        return request.getHeaders().get(HttpHeaders.AUTHORIZATION).get(0).substring("Bearer ".length());
    }

    public String resolveRefreshToken(ServerHttpRequest request){
        return request.getHeaders().get("Refresh").get(0);
    }

    public long getMemberId(String jwtToken){
        return Long.parseLong(extractClaims(jwtToken).getSubject());
    }

    public boolean isTokenValid(String token) {
        final Date expiry = extractClaims(token).getExpiration();
        return expiry.after(new Date());
    }

    private Claims extractClaims(String jwtToken) {
        try {
            return Jwts.parserBuilder()
                    .setSigningKey(secretKey)
                    .build()
                    .parseClaimsJws(jwtToken)
                    .getBody();
        } catch (JwtException e) {
            throw new IllegalStateException();
        }
    }

}

File: backend/tiggle-root/tiggle/src/main/kotlin/com/side/tiggle/global/auth/JwtTokenProvider.kt
package com.side.tiggle.global.auth

import com.side.tiggle.domain.member.repository.MemberRepository
import io.jsonwebtoken.Claims
import io.jsonwebtoken.JwtException
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import org.springframework.stereotype.Component
import java.nio.charset.StandardCharsets
import java.time.LocalDateTime
import java.time.ZoneId
import java.util.*
import jakarta.servlet.http.HttpServletRequest
import javax.crypto.SecretKey

@Component
class JwtTokenProvider(
    private val memberRepository: MemberRepository
) {

    // TODO: @Value로 받아야 한다
    private val secret = "secrettigglesecrettigglesecrettiggle"
    private val secretKey: SecretKey = Keys.hmacShaKeyFor(secret.toByteArray(StandardCharsets.UTF_8))

    private val tokenExpiry = 60L * 30L
    private val refreshTokenExpiry = 60L * 60L * 24L

    private val zone = ZoneId.systemDefault()

    fun generateToken(
        memberId: Long,
        roles: String,
        validFor: Long
    ): String {
        val issued = Date.from(LocalDateTime.now().atZone(zone).toInstant())
        val expiry = Date.from(LocalDateTime.now().plusSeconds(validFor).atZone(zone).toInstant())

        return Jwts.builder()
            .issuer("tiggle")
            .subject(memberId.toString())
            .claim("roles", roles)
            .issuedAt(issued)
            .expiration(expiry)
            .signWith(secretKey)
            .compact()
    }

    fun getAccessToken(
        memberId: Long,
        roles: String
    ): String {
        return generateToken(memberId, roles, tokenExpiry)
    }

    fun getRefreshToken(
        memberId: Long,
        roles: String
    ): String {
        return generateToken(memberId, roles, refreshTokenExpiry)
    }

    fun resolveAccessToken(request: HttpServletRequest): String {
        return request.getHeader("Authorization").substring("Bearer ".length)
    }

    fun resolveRefreshToken(request: HttpServletRequest): String {
        return request.getHeader("Refresh")
    }

    fun getUserId(token: String): Long {
        return extractClaims(token).subject.toLong()
    }

    fun isTokenValid(token: String): Boolean {
        return try {
            val expiry: Date = extractClaims(token).expiration
            expiry.after(Date())
        } catch (e: Exception) {
            println("토큰 검증 실패: ${e.message}")
            false
        }
    }

    private fun extractClaims(
        jwtToken: String
    ): Claims {
        try {
            return Jwts.parser()
                .verifyWith(secretKey)
                .build()
                .parseSignedClaims(jwtToken)
                .payload
        } catch (e: JwtException) {
            throw IllegalStateException()
        }
    }

}